### PR TITLE
Fix MindtPy bugs

### PR DESCRIPTION
**Pull request by ZedongPeng**
## Fixes # .
1. [fix gurobi single tree termination check bug](https://github.com/Pyomo/pyomo/commit/905503b13907a610a1e7f9d8620ee88127551ec5)
2. [fix Gurobi single tree cycle handling](https://github.com/Pyomo/pyomo/commit/3b01104f586f9a33d1386ff89c5c3e57b3ae7fc7)
3. [improve copy_var_list_values function](https://github.com/Pyomo/pyomo/commit/a66f955500686fef09d8605571ef6f5c238101f8)
4. [fix bug in feasibility pump method](https://github.com/Pyomo/pyomo/commit/cefd4a66a06711e90005d20cd470efca762754a6)
5. [add special handle for infeasible relaxed NLP](https://github.com/Pyomo/pyomo/commit/a767f281acd08629aea99f6c616e59b11372d2f5) 
6. [update the log format of infeasible fixed NLP subproblems](https://github.com/Pyomo/pyomo/commit/f766c0a9d6f66da8fea8b11f277bb719ef94f48f)
7. [create new copy_var_list_values function](https://github.com/Pyomo/pyomo/commit/a6e92c53e63b9febad79e8a19a5230c8c308328a)

## Summary/Motivation:
This PR fixes several bugs in MindtPy according to the benchmark results.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
9. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by blnicho** [comment]
This is waiting for #2988 to be merged first

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (emma58) [inline pyomo/contrib/mindtpy/algorithm_base_class.py:860]: Suggestion: change 'try' to 'trying'

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (emma58) [inline pyomo/contrib/mindtpy/algorithm_base_class.py:1687]: `dir()` returns a list, so this `if` statement is linear time, which isn't ideal. Is there a more efficient way to check that you have MIP results?

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (emma58) [inline pyomo/contrib/mindtpy/algorithm_base_class.py:815]: Why not just store the number of OA cuts for each of these solutions? It seems inefficient to store a list of indices when you can always construct a generator from one number, but maybe I'm wrong?

File: pyomo/contrib/mindtpy/util.py
Comment (emma58) [inline pyomo/contrib/mindtpy/util.py:1013]: It looks like this code is copied three times in this PR: Here, in the single-tree implementation and above for the solution pools. Could this part be a utility function to reduce the redundant code?

Comment (emma58) [inline pyomo/contrib/mindtpy/util.py:1013]: It will be helpful to add which Vars you were copying from and to in this error message.

**Review by emma58 — COMMENTED** [review]
A few comments and questions. I also think it would be good to add tests for the bugs you are addressing in this PR.

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:815]: `int_sol_2_cuts_ind` is a `dict`, with `self.curr_int_sol` list as the `key` and `index list` as the `value`. 
Do you want me to improve the definition of the key or the value? I think improving the value is doable.

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:1687]: Is it better to use `locals()`, which returns a dict?

Comment (emma58) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:1687]: If you're going to use `in` then a dict would be much better, yes.

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (emma58) [inline pyomo/contrib/mindtpy/algorithm_base_class.py:1105]: In this pattern, it is still good to include the text from the original error. Here you are hiding it from the user completely and replacing it with your own.

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:815]: That's a good point. However, one integer is not enough. We need two integers denoting the starting index and ending index respectively. I will try to implement it.

**Comment by mrmundt** [comment]
@ZedongPeng - the `gurobi` failures are not your fault. #3053 is going to temporarily fix them; Michael is going to (more) permanently fix them soon.

File: pyomo/contrib/mindtpy/extended_cutting_plane.py
Comment (ZedongPeng) [inline pyomo/contrib/mindtpy/extended_cutting_plane.py:144]: yes

**Comment by ZedongPeng** [comment]
Most of the uncovered lines are `logging.errors`. I added the unit tests for new `set_var_value` function.

**Comment by ZedongPeng** [comment]
I think this PR is ready for a second review.

**Review by emma58 — COMMENTED** [review]
A few more comments, and it would still be best if you could increase coverage: You can turn up the logging level and test some of those error messages.

File: pyomo/contrib/mindtpy/util.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/util.py:999]: I think both docstring and plain comments will work.

Comment (emma58) [reply pyomo/contrib/mindtpy/util.py:999]: I'm just doesn't feel like documentation to me--it seems more like a comment. It's also not clear what "the following" is referring to anymore since there's no longer a try-except pattern.

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:1689]: I tried. However, it's not easy to find a problem to generate a model with an unbounded MIP problem, with the initialization strategy. Tbh, I forgot which instance resulted in this situation in the benchmark tests.

File: pyomo/repn/tests/ampl/test_nlv2.py
Comment (ZedongPeng) [reply pyomo/repn/tests/ampl/test_nlv2.py:1091]: This came from merging the `main` branch into the current `benchmark` branch. It is a little bit wired. Is it good to manually remove this?

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (emma58) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:1689]: You could always make the whole problem unbounded--then you would be sure to hit this (as long as you didn't discover it was unbounded during initialization I guess). I think there are a couple small examples that might be useful to build off of in `test_gdpopt.py`. Another option would be to just directly call the `solve_main` method and skip initialization.

Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:1689]: This code will not be triggered for all unbounded problems. It only happens when the problem is unbounded and the heuristic method in the solver also finds a feasible solution, **as follows**. That's why the problem is unbounded and still calls the `load_solutions` function. I can manually reproduce this situation by changing the code of MindtPy.  However, I didn't find an instance that could result in this situation without any change. It might not be worth to introduce some redundant code just to trigger this error. I will try to increase the code coverage in other parts.
```
Gurobi Optimizer version 10.0.2 build v10.0.2rc0 (mac64[x86])

CPU model: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Thread count: 6 physical cores, 12 logical processors, using up to 12 threads

Optimize a model with 4 rows, 6 columns and 8 nonzeros
Model fingerprint: 0xb497cb3e
Variable types: 5 continuous, 1 integer (0 binary)
Coefficient statistics:
  Matrix range     [1e+00, 4e+00]
  Objective range  [1e+00, 1e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+00, 2e+01]
Found heuristic solution: objective -3.60103e+09
Presolve removed 3 rows and 3 columns
Presolve time: 0.00s

Explored 0 nodes (0 simplex iterations) in 0.00 seconds (0.00 work units)
Thread count was 1 (of 12 available processors)

Solution count 1: -3.60103e+09 
No other solutions better than 0

Model is unbounded
Best objective -3.601030018124e+09, best bound -, gap -
```

**Comment by bernalde** [comment]
All test failures correspond to python >= 3.10 and the GDPOpt LBB test when importing Z3. This is unrelated to this PR but might be good to look into @emma58

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (bernalde) [inline pyomo/contrib/mindtpy/algorithm_base_class.py:816]: The name of this function is unclear

File: pyomo/contrib/mindtpy/single_tree.py
Comment (bernalde) [inline pyomo/contrib/mindtpy/single_tree.py:277]: Should we address this now in this PR or leave it for later?

File: pyomo/contrib/mindtpy/single_tree.py
Comment (bernalde) [inline pyomo/contrib/mindtpy/single_tree.py:681]: This comment is to mention some reasons why we trigger tests?

File: pyomo/contrib/mindtpy/outer_approximation.py
Comment (bernalde) [inline pyomo/contrib/mindtpy/outer_approximation.py:97]: We use """ """ for function documentation

**Review by bernalde — APPROVED** [review]
Very small cosmetic changes in this request. Maybe address to restart tests as previous failures were unrelated to this PR. Use the chance to increase coverage if easy to do.

File: pyomo/contrib/mindtpy/single_tree.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/single_tree.py:681]: To be more precise, the algorithm will terminate during the MIP start process as soon as the bounds converge. In such scenarios, we utilize `self.abort()` to terminate CPLEX. As a consequence, the `results.solver.status` will be set to `error`, accompanied by an error message. Our initial choice to use `self.abort()` was driven by the need for immediate termination upon meeting the convergence criteria. However, I am currently uncertain about the extent of the difference it would make if we allowed CPLEX to terminate on its own.

Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/single_tree.py:681]: Since the error will be captured by the `try except` command. We don't need to worry about it and it is still the most efficient way to terminate CPLEX. I add a TODO here. We can decide based on the benchmark results in the future.

File: pyomo/contrib/mindtpy/single_tree.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/single_tree.py:277]: This has already been checked in the following code. We can remove the TODO.

**Comment by ZedongPeng** [comment]
Could you help me to check why `test_ECP_add_slack` failed only in `linux/pypy-39`? @emma58 @jsiirola
I cannot find the detailed log and the test succeeds in all other environments and my local environment.

**Comment by mrmundt** [comment]
@ZedongPeng - there are issues in general with pypy right now. #3087 is going to fix them.

**Comment by ZedongPeng** [comment]
Hi @mrmundt,

I've noticed that the `pypy-3.9` test continues to fail even after merging #3087. Interestingly, when I ran the tests locally, they all passed successfully. Could you advise me on how to investigate the reasons behind the test exceeding the `timelimit`?

Thank you for your assistance.

**Comment by bernalde** [comment]
The same error seems to be happening although merging master. It always times out in the test_feaspump code from MindtPy, suggesting we are the ones introducing issues here. We need to replicate the testing environment locally to check what's going on. Can we replicate it (in the scenario we can get our hands on a Linux machine)? Somehow, installing promo and some packages via PyPy in Python 3.9 might suggest it is the way forward. Some indication might be useful.

**Comment by ZedongPeng** [comment]
The `osx/3.9` test encountered a failure, which appears to be due to an internet connectivity issue. The specific error message was: `Error: nodename nor servname provided, or not known (api.github.com:443)`. 
Based on the overall progress, I believe this PR is now ready for merging. Could you please give it a final review, @emma58?

File: pyomo/contrib/mindtpy/global_outer_approximation.py
Comment (emma58) [inline pyomo/contrib/mindtpy/global_outer_approximation.py:112]: How can this fail? Wouldn't this be a `DeveloperError` (if you need an error here at all) since mindtpy can keep track of what cuts it has added?

**Review by emma58 — COMMENTED** [review]
A few questions, but overall this looks good. Thanks for the added tests!

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:112]: Thanks for pointing it out. The type of integer solution is `tuple` instead of `list`.

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:1293]: Nope.

File: pyomo/contrib/mindtpy/global_outer_approximation.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/global_outer_approximation.py:112]: Thanks for pointing this out. I checked the log of the benchmark tests and found this error came out when `timelimit` was reached. Let me try to fix it before merging this PR. Thanks.

Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/global_outer_approximation.py:112]: The `KeyError` happens when no feasible solution has been found within the `timelimit`. I added one more condition for the `fix_dual_bound` function in the **OA, GOA, LP/NLP B&B and GLP/NLP B&B** methods. 
I also removed `fix_dual_bound` for the **ECP** method since we will never add no-good cuts in the **ECP** method.
I would appreciate it if @bernalde can give it a double-check.

**Comment by ZedongPeng** [comment]
I have already run the black format locally. Are there any new requirements for the black format?
```
~/Github/pyomo/pyomo/contrib/mindtpy benchmark *1 !2 ❯ black -S -C ./
All done! ✨ 🍰 ✨
40 files left unchanged.
```

**Comment by mrmundt** [comment]
@ZedongPeng - check your black version. They released an update this month that made some changes.

**Review by emma58 — APPROVED** [review]

File: pyomo/contrib/mindtpy/util.py
Comment (blnicho) [inline pyomo/contrib/mindtpy/util.py:947]:
```suggestion
        The variables that provide the values to copy from.
```

File: pyomo/contrib/mindtpy/util.py
Comment (blnicho) [inline pyomo/contrib/mindtpy/util.py:1027]:
```suggestion
            "set_var_valid_value failed with variable {}, value = {} and rounded value = {}"
```

**Review by blnicho — APPROVED** [review]
@ZedongPeng I have a couple minor edits but otherwise I think this looks good.

**Pull request merged**
